src/php/types/array.rs

```diff
@@ -277,202 +198,299 @@ impl<'a> Debug for ZendHashTable<'a> {
+//// HashMap
+///////////////////////////////////////////
+
+impl<V> TryFrom<&HashTable> for HashMap<String, V>
 where
-    V: FromZval<'a>,
+    for<'a> V: FromZval<'a>,
 {
     type Error = Error;
 
-    fn try_from(zht: ZendHashTable<'a>) -> Result<Self> {
-        let mut hm = HashMap::with_capacity(zht.len());
+    fn try_from(value: &HashTable) -> Result<Self> {
+        let mut hm = HashMap::with_capacity(value.len());
 
-        for (idx, key, val) in zht.into_iter() {
+        for (idx, key, val) in value.iter() {
             hm.insert(
                 key.unwrap_or_else(|| idx.to_string()),
-                V::from_zval(val).ok_or(Error::ZvalConversion(val.get_type()?))?,
+                V::from_zval(val).ok_or_else(|| Error::ZvalConversion(val.get_type()))?,
             );
         }
 
         Ok(hm)
     }
 }
 
-impl<'a, K, V> TryFrom<HashMap<K, V>> for ZendHashTable<'a>
+impl<K, V> TryFrom<HashMap<K, V>> for OwnedHashTable
 where
     K: AsRef<str>,
     V: IntoZval,
 {
     type Error = Error;
 
-    fn try_from(hm: HashMap<K, V>) -> Result<Self> {
-        let mut ht =
-            ZendHashTable::with_capacity(hm.len().try_into().map_err(|_| Error::IntegerOverflow)?);
+    fn try_from(value: HashMap<K, V>) -> Result<Self> {
+        let mut ht = OwnedHashTable::with_capacity(
+            value.len().try_into().map_err(|_| Error::IntegerOverflow)?,
+        );
 
-        for (k, v) in hm.into_iter() {
+        for (k, v) in value.into_iter() {
             ht.insert(k.as_ref(), v)?;
         }
 
         Ok(ht)
     }
 }
 
-/// Implementation converting a Rust HashTable into a ZendHashTable.
-impl<'a, 'b, K, V> TryFrom<&'a HashMap<K, V>> for ZendHashTable<'b>
+impl<K, V> IntoZval for HashMap<K, V>
 where
     K: AsRef<str>,
-    V: IntoZval + Clone,
+    V: IntoZval,
 {
-    type Error = Error;
+    const TYPE: DataType = DataType::Array;
 
-    fn try_from(hm: &'a HashMap<K, V>) -> Result<Self> {
-        let mut ht =
-            ZendHashTable::with_capacity(hm.len().try_into().map_err(|_| Error::IntegerOverflow)?);
+    fn set_zval(self, zv: &mut Zval, _: bool) -> Result<()> {
+        let arr = self.try_into()?;
+        zv.set_hashtable(arr);
+        Ok(())
+    }
+}
 
-        for (k, v) in hm.iter() {
-            ht.insert(k.as_ref(), v.clone())?;
-        }
+impl<T> FromZval<'_> for HashMap<String, T>
+where
+    for<'a> T: FromZval<'a>,
+{
+    const TYPE: DataType = DataType::Array;
 
-        Ok(ht)
+    fn from_zval(zval: &Zval) -> Option<Self> {
+        zval.array().and_then(|arr| arr.try_into().ok())
     }
 }
 
-/// Implementation for converting a reference to `ZendHashTable` into a `Vec` of given type.
-/// Will return an error type if one of the values inside the array cannot be converted into
-/// a type `T`.
-impl<'a, V> TryFrom<ZendHashTable<'a>> for Vec<V>
+///////////////////////////////////////////
+//// Vec
+///////////////////////////////////////////
+
+impl<T> TryFrom<&HashTable> for Vec<T>
```

Comment:
Could we get array with `FromIterator` for this? It would be a bit more powerful as you can collect into either `Result<Vec<T>>` or Vec<Result<T>>`:
```rust
let vec: Result<Vec<u32>> = hashmap.iter()
    .map(|(_, _, val)| val.extract()
    .collect();
```

Reply:
Actually, its probably better to just provide more iterators:
```rust
let vec: Result<Vec<u32>> = hashmap.iter_values()
    .map(Zval::extract)
    .collect();
```
And this probably is more flexible as well, as PHP datastructures are not guaranteed to be heterogeneous in the first place.

Reply:
Yeah, iterator is probably better, I only added the impls for ease of use and this way preallocates using `with_capacity`, although now thinking about it the `FromIterator` implementations probably use the `count` function on `Iterator`.

Good idea about the `values()` iterator, will add  it now.

Reply:
My understand is it may if there's a size hint provided.

But it's probably best to leave it to the target collection to decide as well.

Reply:
Do you think these implementations should be removed in favour of iterators? I will also implement `ExactSizeIterator` and `DoubleEndedIterator` for `HashTable`